### Asim Raza

#### Machine Learning

### Import Libraries

In [212]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

### Load dataset

In [213]:
df = pd.read_csv("pizza_v1.csv")

### Print data

In [214]:
df

company price_rupiah  diameter      topping           variant     size  \
0         A    Rp235,000      22.0      chicken  double_signature    jumbo   
1         A    Rp198,000      20.0    papperoni  double_signature    jumbo   
2         A    Rp120,000      16.0    mushrooms  double_signature  reguler   
3         A    Rp155,000      14.0  smoked beef  double_signature  reguler   
4         A    Rp248,000      18.0   mozzarella  double_signature    jumbo   
..      ...          ...       ...          ...               ...      ...   
124       E     Rp39,000       8.5         tuna        spicy tuna    small   
125       E     Rp72,000      12.0         tuna        spicy tuna   medium   
126       E     Rp99,000      14.0         tuna        spicy tuna    large   
127       E     Rp44,000       8.5         meat   BBQ_meat_fiesta    small   
128       E     Rp78,000      12.0         meat   BBQ_meat_fiesta   medium   

    extra_sauce extra_cheese  
0           yes          yes  
1           yes          yes  
2           yes          yes  
3           yes           no  
4           yes           no  
..          ...          ...  
124         yes          yes  
125         yes          yes  
126         yes          yes  
127         yes           no  
128          no           no  

[129 rows x 8 columns]

### Information of data

In [215]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129 entries, 0 to 128
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   company       129 non-null    object 
 1   price_rupiah  129 non-null    object 
 2   diameter      129 non-null    float64
 3   topping       129 non-null    object 
 4   variant       129 non-null    object 
 5   size          129 non-null    object 
 6   extra_sauce   129 non-null    object 
 7   extra_cheese  129 non-null    object 
dtypes: float64(1), object(7)
memory usage: 8.2+ KB


In [216]:
df.shape

(129, 8)

In [217]:
df.isnull().sum()

company         0
price_rupiah    0
diameter        0
topping         0
variant         0
size            0
extra_sauce     0
extra_cheese    0
dtype: int64

In [218]:
df.duplicated().sum()

3

In [219]:
df.describe()

diameter
count  129.000000
mean    12.976744
std      3.272674
min      8.000000
25%     12.000000
50%     12.000000
75%     14.000000
max     22.000000

### Data Cleaning

In [220]:
df['price_rupiah'].unique()

array(['Rp235,000', 'Rp198,000', 'Rp120,000', 'Rp155,000', 'Rp248,000',
       'Rp140,000', 'Rp110,000', 'Rp70,000', 'Rp90,000', 'Rp230,000',
       'Rp188,000', 'Rp114,000', 'Rp149,000', 'Rp23,500', 'Rp46,000',
       'Rp72,000', 'Rp49,000', 'Rp83,000', 'Rp96,000', 'Rp31,000',
       'Rp69,000', 'Rp93,000', 'Rp75,000', 'Rp115,000', 'Rp123,000',
       'Rp33,000', 'Rp76,000', 'Rp119,000', 'Rp126,500', 'Rp39,000',
       'Rp99,000', 'Rp44,000', 'Rp78,000', 'Rp105,000', 'Rp35,000',
       'Rp60,000', 'Rp98,000', 'Rp28,000', 'Rp51,000', 'Rp84,000',
       'Rp32,000', 'Rp54,000', 'Rp92,000'], dtype=object)

In [221]:
df['price_rupiah'] = df['price_rupiah'].str.replace('Rp', '').str.replace(',', '').astype(float)

In [222]:
df

company  price_rupiah  diameter      topping           variant     size  \
0         A      235000.0      22.0      chicken  double_signature    jumbo   
1         A      198000.0      20.0    papperoni  double_signature    jumbo   
2         A      120000.0      16.0    mushrooms  double_signature  reguler   
3         A      155000.0      14.0  smoked beef  double_signature  reguler   
4         A      248000.0      18.0   mozzarella  double_signature    jumbo   
..      ...           ...       ...          ...               ...      ...   
124       E       39000.0       8.5         tuna        spicy tuna    small   
125       E       72000.0      12.0         tuna        spicy tuna   medium   
126       E       99000.0      14.0         tuna        spicy tuna    large   
127       E       44000.0       8.5         meat   BBQ_meat_fiesta    small   
128       E       78000.0      12.0         meat   BBQ_meat_fiesta   medium   

    extra_sauce extra_cheese  
0           yes          yes  
1           yes          yes  
2           yes          yes  
3           yes           no  
4           yes           no  
..          ...          ...  
124         yes          yes  
125         yes          yes  
126         yes          yes  
127         yes           no  
128          no           no  

[129 rows x 8 columns]

In [223]:
df['size'].unique()

array(['jumbo', 'reguler', 'small', 'medium', 'large', 'XL'], dtype=object)

In [224]:
df['size'].replace('XL','Extra_large', inplace = True)

In [225]:
df['size'].unique()

array(['jumbo', 'reguler', 'small', 'medium', 'large', 'Extra_large'],
      dtype=object)

In [226]:
df.head()

company  price_rupiah  diameter      topping           variant     size  \
0       A      235000.0      22.0      chicken  double_signature    jumbo   
1       A      198000.0      20.0    papperoni  double_signature    jumbo   
2       A      120000.0      16.0    mushrooms  double_signature  reguler   
3       A      155000.0      14.0  smoked beef  double_signature  reguler   
4       A      248000.0      18.0   mozzarella  double_signature    jumbo   

  extra_sauce extra_cheese  
0         yes          yes  
1         yes          yes  
2         yes          yes  
3         yes           no  
4         yes           no

In [227]:
df.drop(columns=['company', 'diameter', 'topping', 'variant', 'extra_sauce', 'extra_cheese'], inplace=True)


In [228]:
df

price_rupiah     size
0        235000.0    jumbo
1        198000.0    jumbo
2        120000.0  reguler
3        155000.0  reguler
4        248000.0    jumbo
..            ...      ...
124       39000.0    small
125       72000.0   medium
126       99000.0    large
127       44000.0    small
128       78000.0   medium

[129 rows x 2 columns]

In [229]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df['size'] = le.fit_transform(df['size'])

In [230]:
df

price_rupiah  size
0        235000.0     1
1        198000.0     1
2        120000.0     4
3        155000.0     4
4        248000.0     1
..            ...   ...
124       39000.0     5
125       72000.0     3
126       99000.0     2
127       44000.0     5
128       78000.0     3

[129 rows x 2 columns]

In [231]:
from sklearn.preprocessing import StandardScaler


In [232]:
scaler = StandardScaler()

df['price_rupiah'] = scaler.fit_transform(df[['price_rupiah']])

In [233]:
df

price_rupiah  size
0        3.320022     1
1        2.489168     1
2        0.737638     4
3        1.523581     4
4        3.611944     1
..            ...   ...
124     -1.081259     5
125     -0.340227     3
126      0.266072     2
127     -0.968982     5
128     -0.205494     3

[129 rows x 2 columns]

In [234]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [235]:
x = df[['price_rupiah']]
y = df[['size']]

In [244]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0, shuffle=True)
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

(103, 1) (103, 1)
(26, 1) (26, 1)


In [245]:
linear_regression_model = LinearRegression()

In [246]:
linear_regression_model.fit(x_train, y_train)

LinearRegression()

In [247]:
# predict the target on train and test data
y_pred = linear_regression_model.predict(x_test)
print(y_pred)

[[3.10898934]
 [1.72075653]
 [3.24210755]
 [2.67160092]
 [2.84275291]
 [2.97587112]
 [3.67949597]
 [2.97587112]
 [3.88868173]
 [3.22309066]
 [2.69061781]
 [1.89190852]
 [3.67949597]
 [2.78570224]
 [2.72865158]
 [2.78570224]
 [2.27224627]
 [3.41325954]
 [3.18505689]
 [3.67949597]
 [2.3863476 ]
 [3.12800622]
 [2.72865158]
 [3.58441153]
 [3.62244531]
 [3.81261418]]


In [250]:
score_r2 = r2_score(y_test, y_pred)
print(score_r2*100)

0.1528994391616023


In [254]:
from sklearn.linear_model import ARDRegression
from sklearn.metrics import r2_score

model = ARDRegression()

model.fit(x_train, y_train)

y_pred = model.predict(x_test)
print(y_pred)


[3.10457155 1.7413801  3.23528853 2.67507287 2.84313757 2.97385456
 3.66478721 2.97385456 3.87019962 3.21661468 2.69374673 1.9094448
 3.66478721 2.787116   2.73109444 2.787116   2.28292191 3.40335323
 3.17926697 3.66478721 2.39496504 3.1232454  2.73109444 3.57141793
 3.60876564 3.7955042 ]


c:\Users\h\anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [255]:
score_r2 = r2_score(y_test, y_pred)
print(score_r2*100)

0.5536484910829476


In [256]:
from sklearn.linear_model import HuberRegressor
from sklearn.metrics import r2_score

model = HuberRegressor()

model.fit(x_train, y_train)

y_pred = model.predict(x_test)
print(y_pred)

[3.17837098 1.76018977 3.31436096 2.73154676 2.90639102 3.042381
 3.76118517 3.042381   3.97488371 3.29493382 2.7509739  1.93503403
 3.76118517 2.8481096  2.78982818 2.8481096  2.32357682 3.48920521
 3.25607954 3.76118517 2.44013966 3.19779812 2.78982818 3.66404947
 3.70290375 3.89717515]


c:\Users\h\anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [257]:
score_r2 = r2_score(y_test, y_pred)
print(score_r2*100)

-1.88093249773309


In [261]:
from sklearn.linear_model import PassiveAggressiveRegressor
from sklearn.metrics import r2_score

model = PassiveAggressiveRegressor()

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

print(y_pred)

[2.254003   0.36362733 2.4352719  1.65840519 1.8914652  2.0727341
 3.03086972 2.0727341  3.31572085 2.40937635 1.68430075 0.59668734
 3.03086972 1.81377853 1.73609186 1.81377853 1.11459849 2.66833192
 2.35758523 3.03086972 1.26997183 2.27989856 1.73609186 2.90139193
 2.95318305 3.21213862]


c:\Users\h\anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [260]:
score_r2 = r2_score(y_test, y_pred)
print(score_r2*100)

-1.9627004603417708


### conclusion 

### Here efficency of  ARDRegression is Best about 55%
### Because many outliers is Present